### PR TITLE
Create table to store api tokens for the discord bot

### PR DESCRIPTION
**Pull request by topocount**
The schema and permissions support the following flow:

1. The discord bot receives a request to administer a circle via the
   discord bot. This might entail the following:
   - adding or removing users via discord roles
   - assigning the circle to a channel
   - configure circle notifications in discord
2. the bot inserts a row in the discord.circle_api_tokens table
   containing a `circle_id` and the `channel_snowflake` and begins listening for the token to be added
   to that row via a subscription
3. the bot generates a link that will open the coordinape app on the
   discord route with the param `id` representing the row id in the
   circle_api_tokens table.
4. The user opens the link in a browser and authenticates or is
   authenticated with the address that has admin perms on the circle.
5. In the frontend, a circle api token is generated and passed into an
   action handler along with the circle_id provided, and the id of the
   token.
6. The handler attempts to update the row created by the bot with the
   new id. If no row exists with a null token field and the
   update fails, the key is deleted and an error is shown to the user
   in a toast on redirect.
7. If the update is successful, the user is notified the circle is
   linked to the channel in both the app in a toast and in discord
   because the subscription returns affirmatively.

We want to ensure the links generated to complete the flow are
single-use and do not cause unintended side-effects if the link is
accessed more than once. This flow guarantees that the links will be
single-use.

This schema and prescribed flow would not be possible for updating the
api token, so it's best to request all permissions the bot needs up
front, rather than having the permissions granted to the bot evolve
without compromising on the single-use aspect.

We could consider generating and passing around a single-use identifier
but the flow gets a bit more complex. This could also be added later
without much concern for additional lift. [YAGNI](https://builtin.com/software-engineering-perspectives/yagni) for now seems prudent
though

### Todos
- [ ] Update the Zeus schema

### CODE REVIEW

**Comment by topocount** [comment]
Current dependencies on/for this PR:
* main
  * **PR #1740** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1740" target="_blank">Graphite</a> 
    * **PR #1741** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1741" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/coordinape/coordinape/1741?utm_source=stack-comment).

**Comment by topocount** [comment]
TODO: update the new zeus schema

**Review by levity — DISMISSED** [review]

**Comment by karelianpie** [comment]
I'm not sure about the steps 2-6, I thought after out discussion that the flow would be

1. The bot generates a link that will open the coordinape app on the discord route with the params `circle_id` and `channel_snowflake`;
2. The user clicks on the link and is redirected from discord to the browser;
3. In the frontend, a circle api token is generated and passed into an action handler along with the circle_id provided, and the id of the token;
4. The handler attempts to insert a new row, if there's already a row with `circle_id` and `channel_snowflake` then an error message is displayed letting the user know that the circle is already linked to the channel. If there is no row with the given `circle_id` and `channel_snowflake` and the insert is successful, the user is notified the circle is linked to the channel in both the app in a toast and in discord.

I'm not sure why we need the subscription, can't this be done synchronously through a mutation?

I'm afraid that we will also need the user to authorise discord yet again in this step too. The reason being that we need the discord admin user snowflake as well, this should not be sent as a param otherwise anyone could try and impersonate a circle admin. Therefore, I believe we will need to have another user authorisation, once we have the user token we can make a request to the discord api to get their snowflake as we do for the user linking, once we have their snowflake we can initiate this flow because we can validate that the user making the request is indeed the discord user linked to the profile in coordinape that as admin permission to the circle.

> This schema and prescribed flow would not be possible for updating the
api token, so it's best to request all permissions the bot needs up
front, rather than having the permissions granted to the bot evolve
without compromising on the single-use aspect.

This is a bit of a concern as well, unless we know all permissions that we need upfront. What if new things are added to the circle, how would we update the tokens?

**Comment by topocount** [comment]
@karelianpie the flow you describe above makes the generated links non-idempotent, and will generate another token each time the link is accessed in an authenticated context. We want the links to be strictly single-use.

> What if new things are added to the circle, how would we update the tokens?

If the discord bot doesnt have the perms, it can always request more. We'd just pass the existing row `id` into an update handler that deletes the exiting row in a database transaction that subsequently adds the new one. I'm trying to do you a favor here by making it plain that this concern is not in scope for your work :wink: but if you want to do it, I'm not going to protest.

My point was that this flow cannot be used to create new token unless the existing one is explicitly unlinked first. This is because the existing row would throw a constraint error, since circle_ids and tokens have a 1:1 relationship in this table. (to minimize ambiguity)

> I'm not sure why we need the subscription, can't this be done synchronously through a mutation?

the subscription is needed only for the bot to listen for a token to be added for a given circle. Apologies if my wording was confusing there.

> I'm afraid that we will also need the user to authorize discord yet again in this step too. The reason being that we need the discord admin user snowflake as well.

This won't be necessary, because as @levity suggested, the discord_bot role will be updated have access to user perm levels and can verify a linked coordinape user's circle role before proceeding to generate a link.

Please let me know if you have any further questions.

**Review by karelianpie — APPROVED** [review]

**Comment by karelianpie** [comment]
Can this be merged @topocount?

**Comment by karelianpie** [comment]
Ok, so in the discord front:

1. the bot can already request the user to select the circles they want to manage (we still need to get the real circles):

Screenshot 2022-12-20 at 4 01 57 pm

2. Create a new channel and role for the circle(s) selected:

Screenshot 2022-12-20 at 4 03 05 pm

3. In the new channel prompt the user to link the circle, once they click in Link it's ready to send the mutation with `circleId` and `channel_snowflake`:

Screenshot 2022-12-20 at 4 00 07 pm

Once this goes in we can and begin listening for the token to be added to that row via a subscription and continue with the flow

/cc @reeserj

File: hasura/migrations/default/1670795162160_alter_table_discord_roles_circles_alter_column_role/up.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+alter table "discord"."roles_circles" rename column "role" to "server_role";
```
Comment (crabsinger) [inline hasura/migrations/default/1670795162160_alter_table_discord_roles_circles_alter_column_role/up.sql:1]: Can we move this into the first migration where we create the table and remove the second `alter table` migration?

Comment (karelianpie) [reply hasura/migrations/default/1670795162160_alter_table_discord_roles_circles_alter_column_role/up.sql:1]: The first migration in this PR created a different table, `"discord"."circle_api_tokens"`, so we should keep this `alter table` as that table creation happened a while ago.

**Comment by karelianpie** [comment]
> @karelianpie I can help get this merged. There was one remaining comment on the duplicate migrations and it needs to be rebased and cleaned up again. If you can make those changes I'll approve and merge. (If you need to push to a different branch because you don't have write to this branch just lmk.)

I do not have write access to this, but was able to rebase locally without any merge conflicts.

Assuming it's all good, let me know @crabsinger

**Comment by karelianpie** [comment]
Let's close this and continue here https://github.com/coordinape/coordinape/pull/1788

I've added a bit more logic to handle the circle linking

**Comment by karelianpie** [comment]
Closing as I've rebased with `main` and added a bit more logic to handle the circle linking here https://github.com/coordinape/coordinape/pull/1788

**Pull request closed without merging**